File: viewer/TweetViewer/Services/ReadMarkQueue.cs
using Microsoft.Data.Sqlite;
using TweetViewer.Data;

namespace TweetViewer.Services;

/// <summary>
/// スクロール既読化のバッチ書込。1秒毎 or 100件到達で read_state に
/// 1トランザクションで INSERT する。クローズ/ユーザー切替時は FlushAsync を呼ぶこと。
/// </summary>
public sealed class ReadMarkQueue : IAsyncDisposable
{
    private const int FlushThreshold = 100;
    private static readonly TimeSpan FlushInterval = TimeSpan.FromSeconds(1);

    private readonly ViewerDatabase _db;
    private readonly Dictionary<string, string> _pending = new();   // tweet_id -> username
    private readonly object _lock = new();
    private readonly PeriodicTimer _timer;
    private readonly Task _loop;
    private readonly CancellationTokenSource _cts = new();

    public ReadMarkQueue(ViewerDatabase db)
    {
        _db = db;
        _timer = new PeriodicTimer(FlushInterval);
        _loop = Task.Run(FlushLoopAsync);
    }

    public void Enqueue(string tweetId, string username)
    {
        var flushNow = false;
        lock (_lock)
        {
            _pending[tweetId] = username;
            flushNow = _pending.Count >= FlushThreshold;
        }
        if (flushNow)
            _ = FlushAsync();
    }

    private async Task FlushLoopAsync()
    {
        try
        {
            while (await _timer.WaitForNextTickAsync(_cts.Token).ConfigureAwait(false))
                await FlushAsync().ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
        }
    }

    public async Task FlushAsync()
    {
        List<KeyValuePair<string, string>> batch;
        lock (_lock)
        {
            if (_pending.Count == 0)
                return;
            batch = _pending.ToList();
            _pending.Clear();
        }

        await _db.WriteLock.WaitAsync().ConfigureAwait(false);
        try
        {
            await Task.Run(() =>
            {
                using var conn = _db.OpenConnection();
                using var tx = conn.BeginTransaction();
                var now = JsonlImporter.UtcNow();
                foreach (var (tweetId, username) in batch)
                {
                    using var cmd = conn.CreateCommand();
                    cmd.Transaction = tx;
                    cmd.CommandText = """
                        INSERT OR IGNORE INTO read_state (tweet_id, username, read_at)
                        VALUES ($id, $u, $t)
                        """;
                    cmd.Parameters.AddWithValue("$id", tweetId);
                    cmd.Parameters.AddWithValue("$u", username);
                    cmd.Parameters.AddWithValue("$t", now);
                    cmd.ExecuteNonQuery();
                }
                tx.Commit();
            }).ConfigureAwait(false);
        }
        catch (SqliteException)
        {
            // 失敗分は再キュー(次のフラッシュで再試行)
            lock (_lock)
            {
                foreach (var (tweetId, username) in batch)
                    _pending.TryAdd(tweetId, username);
            }
        }
        finally
        {
            _db.WriteLock.Release();
        }
    }

    public async ValueTask DisposeAsync()
    {
        _cts.Cancel();
        _timer.Dispose();
        try
        {
            await _loop.ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
        }
        await FlushAsync().ConfigureAwait(false);
        _cts.Dispose();
    }
}

File: mac/Sources/SubakoApp/Services/ReadMarkQueue.swift
import Foundation
import SubakoCore

/// スクロール既読化のバッチ書込 (C# Services/ReadMarkQueue.cs の移植)。
/// 1 秒毎 or 100 件到達で read_state に 1 トランザクションで INSERT する。
/// 表示対象の切替時・アプリ終了時は flush() を呼ぶこと。
/// 閲覧専用モードでは何も書かない。
actor ReadMarkQueue {
    private static let flushThreshold = 100
    private static let flushInterval: Duration = .seconds(1)

    private let repo: TweetRepository
    private let readOnly: Bool
    private var pending: [String: String] = [:]   // tweet_id -> username
    private var loopTask: Task<Void, Never>?

    init(repo: TweetRepository, readOnly: Bool) {
        self.repo = repo
        self.readOnly = readOnly
    }

    func start() {
        guard loopTask == nil, !readOnly else { return }
        loopTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: Self.flushInterval)
                await flush()
            }
        }
    }

    func enqueue(tweetId: String, username: String) async {
        guard !readOnly else { return }
        pending[tweetId] = username
        if pending.count >= Self.flushThreshold {
            await flush()
        }
    }

    func flush() async {
        guard !pending.isEmpty else { return }
        let batch = pending
        pending = [:]
        do {
            try await repo.markRead(batch.map { (tweetId: $0.key, username: $0.value) })
        } catch {
            // 失敗分は再キュー (次のフラッシュで再試行)
            for (tweetId, username) in batch where pending[tweetId] == nil {
                pending[tweetId] = username
            }
            AppLog.error("既読書込に失敗 (再キュー): \(error)")
        }
    }

    func shutdown() async {
        loopTask?.cancel()
        loopTask = nil
        await flush()
    }
}
